### PR TITLE
prov/efa: Handle recv cancel for zero-copy recv

### PR DESCRIPTION
**Pull request by shijin-aws**
Currently, posted recv is not tracked in zero-copy mode,
which breaks various ep operations including fi_cancel. This patch fixes this
issue by introducing a user_recv_rxe_list that tracks the posted user recv,
and implement fi_cancel operation for this list.

### CODE REVIEW

**Comment by j-xiong** [comment]
@shijin-aws Is this the only remaining PR that needs go into 1.22?

**Comment by shijin-aws** [comment]
@j-xiong yes

**Comment by shijin-aws** [comment]
I may need to update the PR later. I will try to make it ready tonight. If it is too late I can backport it after the release is cut.

**Comment by j-xiong** [comment]
I will wait until 5pm tomorrow (Wednesday).

**Comment by shijin-aws** [comment]
@j-xiong Thanks!

**Review by wenduwan — APPROVED** [review]

**Comment by shijin-aws** [comment]
added unit test, should be the last push. Waiting for CI to pass

File: prov/efa/src/rdm/efa_rdm_ep_fiops.c [hunk 1]
```diff
@@ -1288,6 +1289,60 @@ static int efa_rdm_ep_ctrl(struct fid *fid, int command, void *arg)
...
+ * @param context pointer to the context to be cancelled
+ * @return ssize_t 0 on success, negative integer on failure.
+ */
+static
+ssize_t efa_rdm_ep_cancel_zcpy_recv(struct efa_rdm_ep *ep,
+				     void *context)
+{
+	struct dlist_entry *entry;
+	struct efa_rdm_ope *rxe;
+	struct fi_cq_err_entry err_entry;
```
Comment (darrylabbate) [inline prov/efa/src/rdm/efa_rdm_ep_fiops.c:1312]: Zero-initialize and avoid the `memset()` below.

```suggestion
	struct fi_cq_err_entry err_entry = {0};
```

File: prov/efa/src/rdm/efa_rdm_ep_fiops.c [hunk 2]
```diff
@@ -1288,6 +1289,60 @@ static int efa_rdm_ep_ctrl(struct fid *fid, int command, void *arg)
...
+	rxe = container_of(entry, struct efa_rdm_ope, entry);
+	assert(rxe->user_rx_pkt);
+	rxe->user_rx_pkt->flags |= EFA_RDM_PKE_USER_RECV_CANCEL;
+
+	dlist_remove(&rxe->entry);
+	memset(&err_entry, 0, sizeof(err_entry));
+	err_entry.op_context = rxe->cq_entry.op_context;
+	err_entry.flags |= rxe->cq_entry.flags;
+	err_entry.err = FI_ECANCELED;
+	err_entry.prov_errno = -FI_ECANCELED;
```
Comment (darrylabbate) [inline prov/efa/src/rdm/efa_rdm_ep_fiops.c:1336]: Shouldn't this be an EFA-specific code?

Comment (shijin-aws) [reply prov/efa/src/rdm/efa_rdm_ep_fiops.c:1336]: recv cancel is a general libfabric operation, there is no efa specific thing at all

>fi_cancel attempts to cancel an outstanding asynchronous operation. Canceling an operation causes the fabric provider to search for the operation and, if it is still pending, complete it as having been canceled. An error queue entry will be available in the associated error queue with error code FI_ECANCELED.

File: prov/efa/src/rdm/efa_rdm_ep_fiops.c (hunk 1 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/efa_rdm_ep_fiops.c]: Updated

File: prov/efa/src/rdm/efa_rdm_ep_fiops.c (hunk 2 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/efa_rdm_ep_fiops.c:1336]: I don't think there is something else we can put on prov_errno, I am following https://github.com/ofiwg/libfabric/blob/main/prov/util/src/util_srx.c#L888-L890

**Review by darrylabbate — APPROVED** [review]

**Pull request merged**